script/state_manager.js

```diff
@@ -0,0 +1,440 @@
+/*
+ * Module for handling States
+ *
+ * All states should be get and set through the StateManager ($SM).
+ *
+ * The manager is intended to handle all needed checks and error catching.
+ * This includes creating the parents of layered/deep states so undefined states
+ * do not need to be tested for and created beforehand.
+ *
+ * When a state is changed, an update event is sent out containing the name of the state
+ * changed or in the case of multiple changes (.setM, .addM) the parent class changed.
+ * Event: type: 'stateUpdate', stateName: <path of state or parent state>
+ *
+ * Original file created by: Michael Galusha
+ */
+
+var StateManager = {
+
+	MAX_STORE: 99999999999999,
+
+	options: {},
+
+	init: function(options) {
+		this.options = $.extend(
+				this.options,
+				options
+		);
+
+		//create categories
+		var cats = [
+			'features',     // big features like buildings, location availability, unlocks, etc
+			'stores',       // little stuff, items, weapons, etc
+			'character',    // this is for player's character stats such as perks
+			'income',
+			'timers',
+			'game',         // mostly location related: fire temp, workers, population, world map, etc
+			'playStats',    // anything play related: play time, loads, etc
+			'previous',     // prestige, score, trophies (in future), achievements (again, not yet), etc
+			'outfit',      	// used to temporarily store the items to be taken on the path
+			'config',
+			'wait',			// mysterious wanderers are coming back
+			'cooldown'      // residual values for cooldown buttons
+		];
+
+		for(var which in cats) {
+			if(!$SM.get(cats[which])) $SM.set(cats[which], {});
+		}
+
+		//subscribe to stateUpdates
+		$.Dispatch('stateUpdate').subscribe($SM.handleStateUpdates);
+	},
+
+	//create all parents and then set state
+	createState: function(stateName, value) {
+		var words = stateName.split(/[.\[\]'"]+/);
+		//for some reason there are sometimes empty strings
+		for (var j = 0; j < words.length; j++) {
+			if (words[j] === '') {
+				words.splice(j, 1);
+				j--;
+			}
+		}
+		var obj = State;
+		var w = null;
+		for(var i=0, len=words.length-1;i<len;i++){
+			w = words[i];
+			if(obj[w] === undefined ) obj[w] = {};
+			obj = obj[w];
+		}
+		obj[words[i]] = value;
+		return obj;
+	},
+
+	//set single state
+	//if noEvent is true, the update event won't trigger, useful for setting multiple states first
+	set: function(stateName, value, noEvent) {
```

Comment:
The "set" function along with the "setM" function could be combined as they have very similar functionality. Adding one additional parameter could determine what the functionality of the method, rather than having two separate methods.